src/Symfony/Component/Security/Core/Encoder/SodiumPasswordEncoder.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\Encoder;
+
+use Symfony\Component\Security\Core\Exception\BadCredentialsException;
+use Symfony\Component\Security\Core\Exception\LogicException;
+
+/**
+ * Hashes passwords using libsodium.
+ *
+ * @author Robin Chalas <robin.chalas@gmail.com>
+ * @author Zan Baldwin <hello@zanbaldwin.com>
+ * @author Dominik Müller <dominik.mueller@jkweb.ch>
+ *
+ * @final
+ */
+class SodiumPasswordEncoder extends BasePasswordEncoder implements SelfSaltingEncoderInterface
+{
+    private $memoryCost;
+    private $timeCost;
+    private $threads;
+
+    public function __construct(int $memoryCost = null, int $timeCost = null, int $threads = null)
+    {
+        $this->memoryCost = $memoryCost;
+        $this->timeCost = $timeCost;
+        $this->threads = $threads;
+    }
+
+    public static function isSupported()
+    {
+        if (\defined('PASSWORD_ARGON2I') || \defined('PASSWORD_ARGON2ID')) {
+            return true;
+        }
+
+        if (\class_exists('ParagonIE_Sodium_Compat') && \method_exists('ParagonIE_Sodium_Compat', 'crypto_pwhash_is_available')) {
+            return \ParagonIE_Sodium_Compat::crypto_pwhash_is_available();
+        }
+
+        return \function_exists('sodium_crypto_pwhash_str') || \extension_loaded('libsodium');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function encodePassword($raw, $salt)
+    {
+        if ($this->isPasswordTooLong($raw)) {
+            throw new BadCredentialsException('Invalid password.');
+        }
+
+        if (\defined('PASSWORD_ARGON2I') || \defined('PASSWORD_ARGON2ID')) {
+            return password_hash($raw, \defined('PASSWORD_ARGON2ID') ? \PASSWORD_ARGON2ID : \PASSWORD_ARGON2I, [
```

Comment:
does this make sense in this class?

Reply:
well, we must still specify the algorithm to use for `password_hash()` (and this encoder must use a sodium one). For now these two algorithms are the only sodium ones that PHP support natively (same for sodium itself AFAIK). The plan is to handle new algorithms as time goes by. Makes sense?

Reply:
done